Fix passing escaped double quoted spaces to native executables

douglaswth opened the pull request:
This fixes a regression that seems to have been introduced in PowerShell 5 where spaces in segments of a string with an even number of quotation marks even though some may be escaped are ignored for quoting when executing native programs.

I discovered this issue when trying to run a utility that needs to have double quotes to delimit strings in its query language:

``` powershell
rsc json --x1 "object:has(.name:val(\`"rs low space on C: drive\`"))"
```

With PowerShell versions before 5, this worked correctly and the query argument was accepted, but with PowerShell 5 the argument passed was getting through to the utility broken up: `object:has(.name:val("rs`, `low`, `space`, `on`, `C:`, `drive"))`.

Out of curiosity, I tried it out with the latest PowerShell 6 alpha and the same issue still exists, even on Linux! So I have come up with this fix which is to not count backslash escaped quotes in the quote count for the `appendOneNativeArgument` in `NativeCommandParameterBinder`.

douglaswth committed:
Fix passing escaped double quoted spaces to native executables

msftclas commented:
Hi __@douglaswth__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, MSBOT;

vors commented:
Thank you, @douglaswth !
Can you, please, add tests?

vors commented:
To properly test it, we would need a cross-platform equivalent of [echoargs](https://chocolatey.org/packages/echoargs). It's a simple program that just returns the arguments that been passed.

We can implement it as `dotnet` application, similar to [ResGen](https://github.com/PowerShell/PowerShell/tree/master/src/ResGen) that we use in the build process.

rkeithhill commented:
It's a very simple command.  Feel free to use this - if you want: https://github.com/Pscx/Pscx/blob/master/Src/EchoArgs/EchoArgs.cs

douglaswth committed:
Add test/csharp/csharp.nuget.props to .gitignore

douglaswth committed:
Add tests for native command arguments

douglaswth commented:
@vors I have added tests that use `EchoArgs` via `dotnet`.

@rkeithhill `EchoArgs` didn't quite compile on .NET Core due to the use of `Environment.CommandLine`, but that part was not necessary for my test so I have just left it commented out in the import.

vors commented on .gitignore:
```diff
@@ -5,6 +5,7 @@ project.lock.json
 /debug/
 /staging/
 /Packages/
+test/csharp/csharp.nuget.props
```
Is it needed? What does create this file?

douglaswth replied:
I'm not sure, but it was showing up at least on my Linux box.

vors replied:
Perhaps, we should do a wild-card, like `*.nuget.props`

vors commented on test/EchoArgs/EchoArgs.cs:
```diff
@@ -0,0 +1,28 @@
+                Console.WriteLine("Arg {0} is <{1}>", i, args[i]);
+            }
+
+            /*Console.WriteLine("\nCommand line:");
```
Let's remove this line completely and the trailing new-line as well.

vors commented on test/EchoArgs/EchoArgs.cs:
```diff
@@ -0,0 +1,28 @@
+﻿//---------------------------------------------------------------------
```
Let's move EchoArgs under `test/tools/EchoArgs`

vors commented on test/powershell/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -0,0 +1,31 @@
+    function Start-EchoArgs {
+        Push-Location $echoArgsDir
+        try {
+            dotnet run $args
```
To minimize interference with `dotnet` cli, it's desirable to replace `dotnet run` by `dotnet publish --output <path-to-EchoArgs>/run` and direct native executable call.
And we probably should move this step out from the tests to the `build.psm1`.

This way, tests would not depend directly on`dotnet` and it's presents in PATH

vors commented:
Wow, that was fast! Great job! Let's make it vigorous. 
🕐

douglaswth committed:
Move and "publish" EchoArgs for the tests

* Move EchoArgs from test/EchoArgs to test/tools/EchoArgs
* Use "dotnet publish" for building EchoArgs in build.psm1 so the test
  can call it directly

vors commented on .gitignore:
```diff
@@ -1,10 +1,12 @@
 bin/
 obj/
+run/
```
Sorry for hustle, I know that I suggested this name `run` yesterday. I didn't find a good one at the moment, but now I think we should use `bin` instead of `run`, because it would match dotnet output.

vors commented on build.psm1:
```diff
@@ -561,6 +575,8 @@ function Start-PSPester {
 
     Write-Verbose "Running pester tests at '$path' with tag '$($Tag -join ''', ''')' and ExcludeTag '$($ExcludeTag -join ''', ''')'" -Verbose
+    # Publish EchoArgs so it can be run by tests
+    Publish-EchoArgs
```
I like the way you structured it! One nit-picking: maybe rename it to `Publish-PSTestTools`, so we can use to for similar cases. And move the comment above to the function.

vors commented on test/tools/EchoArgs/EchoArgs.cs:
```diff
@@ -0,0 +1,24 @@
+﻿//---------------------------------------------------------------------
```
Can you, please include url from which is was originally obtained to the header?

vors commented:
:clock230:

douglaswth committed:
Rename Publish-EchoArgs to Publish-PSTestTools

* Rename Publish-EchoArgs to Publish-PSTestTools so it can be used
  for other tools as well in the future
* Publish EchoArgs to the bin directory instead of run to match
  convention
* Add source URL to EchoArgs header comment
* Use wildcard of "*.nuget.props" to match
  "test/csharp/csharp.nuget.props" in .gitignore

vors commented on test/powershell/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -0,0 +1,22 @@
+        & $echoArgs $a 'a"b c"d' a"b c"d >$testPath
+        $testPath | Should Contain 'Arg 0 is <ab cd>'
+        $testPath | Should Contain 'Arg 1 is <ab cd>'
+        $testPath | Should Contain 'Arg 2 is <ab cd>'
```
`Should Contain` usage here makes it's harder to troubleshoot the error. When it accured.
Here, I'm trying to execute the test on the inbox powershell v5.1

```
Describing Native Command Arguments
 [+] Should handle quoted spaces correctly 619ms
 [-] Should handle spaces between escaped quotes 127ms
   Expected: file TestDrive:\test.txt to contain {Arg 0 is <a"b c"d>}
   19:         $testPath | Should Contain 'Arg 0 is <a"b c"d>'
   at <ScriptBlock>, F:\dev\PowerShell\test\powershell\Scripting\NativeExecution\NativeCommandArguments.Tests.ps1: line 19
```

As you see, I don't have information about the produced outpuit immediately in my disposal.
Just to find what we are currently produce I would need to run commands by hands.

I suggest to replace `$testPath | Should Contain 'Arg 0 is <ab cd>'` by

```powershell
$lines = Get-Contetxt $testPath
($lines | measure).Count | Should Be 3  # this line will help get a better error message in case there is only one line in the output
$lines[0] | Should Be 'Arg 0 is <ab cd>'
```

lzybkr replied:
Why bother writing the file?  You can just capture the output in a variable with:
```powershell
$lines = & $echoArgs $a 'a"b c"d' a"b c"d
```

douglaswth replied:
I was trying to do what I thought would be reasonable, but it didn't seem to work with Pester:

```powershell
& $echoArgs $a 'a"b c"d' a"b c"d | Should Match @"
Arg 0 is <ab cd>
Arg 1 is <ab cd>
Arg 2 is <ab cd>
"@
```

It seems like Pester only tries to match the first line of output when using `Should Match`. 😕

vors replied:
It's because powershell take output line by line.
You can add ` |  Out-String | ` to get a single string.
Mulit-line probably indeed appropriate here, but I'd like to bring up couple things:
* please, use `Be`, not `Match`, because `Match` takes regex.
* Multi-line output will likely cause problems on cross-platforms, because of line-endings.

I suggest use what I described above, with array of 3 strings. It avoids problems with line-ends and produce a reasonable good error message.

vors commented on test/powershell/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    It "Should handle spaces between escaped quotes" {
+        & $echoArgs 'a\"b c\"d' "a\`"b c\`"d" >$testPath
```
I'd like to understand this better.

```
"a\`"b c\`"d"
```

There is a mix of two things:
backticks as powershell string escape mechanics and `\` as command line args escape mechanics

This is an equivalent of  `'a\"b c\"d'`
When I use ehcoargs from cmd:

```
EchoArgs.exe a\"b c\"d
```

I'm getting

Arg 0 is <a"b>
Arg 1 is <c"d>

So the behavior that is currently described with `Arg 1` doesn't seem right.

douglaswth replied:
That's because it should be sent to EchoArgs as `EchoArgs.exe "a\"b c\"d"` with the outer quotes.

vors replied:
I see. I'd like to dig this a little bit more on all platforms, just to make sure we are doing the right thing.
At the meantime, can I ask you to add comments in the tests, so it's easier to maintain them?
Particularly the line 

```
"a\`"b c\`"d"
```
 already requires knowledge about both PS and native args passing escaping.

Will come back to this today after some experiments.
Thank you, again, for such thoughtful approach to the PR!

vors commented:
I verified this behavior is aligned with bash and cmd.
We will merge it when you address comments about the tests.

douglaswth committed:
Use a variable instead of a file for testing args

* Use a variable (which can be indexed by line) instead of a file to
  test the output of arguments passed to echoargs
* Add comments explaining the tests

vors commented:
Thank you so much for the great contribution!

vors commented:
I just realised where this `*.nuget.props` file came from:

```
log  : Generating MSBuild file .../test.nuget.props.
```

cc @andschwa

andschwa commented:
Ah, looks like the CLI team is making some serious progress!

be5invis commented:
We still need to add a backslash before a backtick-escaped double quote? This does not solve the double-escaping problem. (That is, we have to escape a double quote for both PowerShell and CommandLineToArgvW.)
